samza-core/src/main/java/org/apache/samza/task/AsyncRunLoop.java

```diff
@@ -611,14 +616,30 @@ private boolean isReady() {
       if (coordinatorRequests.commitRequests().remove(taskName)) {
         needCommit = true;
       }
-      if (needWindow || needCommit || endOfStream) {
-        // ready for window or commit only when no messages are in progress and
-        // no window/commit in flight
+
+      boolean windowOrCommitInFlight = windowInFlight || commitInFlight;
+      /*
+       * A task is ready to commit, when task.commit(needCommit) is requested either by user or commit thread
+       * and either of the following conditions are true.
+       * a) When process, window, commit are not in progress.
+       * b) When task.async.commit is true and window, commit are not in progress.
+       */
+      if (needCommit) {
+        return (messagesInFlight.get() == 0 || isAsyncCommitEnabled) && !windowOrCommitInFlight;
+      } else if (needWindow || endOfStream) {
+        /*
+         * A task is ready for window operation, when task.window(needWindow) is requested by either user or window thread
+         * and window, commit are not in progress.
+         */
         return messagesInFlight.get() == 0 && !windowOrCommitInFlight;
```

Comment:
@nickpan47 : 

If I understand the comment correctly, it's pointed out that the only process calls are allowed to go through when commit is in progress and commit is not allowed to go through when messagesInFlight.get() > 0(process is in progress) which is the main cause for burst in message process rate. 

I think the check on line 628 does that and answers the question. If needCommit is true, if either asyncCommit is enabled and neither window  nor commit is in flight, commit is chosen as next step for the task. By that this will allow the commit to scheduled simultaneously when process calls are in progress. This will reduce the burstiness caused by stopping the commit when messagesInFlight > 0(Point 1. will be accomplished by this.)

-- Point 2: bookkeeping the low watermark of each input SSP's offsets 
TaskCallBackManager(existing implementation) which handles the offsets from the completed callbacks, updates the OffsetManager with the highest sequence number message that is processed & committed by the task. This logic is agnostic of input ssp, amongst all input SSP's highest sequence number  of message that is committed & processed by task is chosen for commit. If I understand this point correctly(picking lowest offset for each input ssp for commit), what you're asking for is done in PR 87.

Sure, let's discuss this in person. Please point out if i'm misunderstanding something.